webapi/setaltsig_test.go

```diff
@@ -0,0 +1,276 @@
+		wantCode:  http.StatusBadRequest,
+	}, {
+		name:      "bad request",
+		deformReq: 1,
+		wantCode:  http.StatusBadRequest,
+	}, {
+		name:     "bad addr",
+		addr:     "xxx",
+		wantCode: http.StatusBadRequest,
+	}, {
+		name:     "addr wrong type",
+		addr:     "DkM3ZigNyiwHrsXRjkDQ8t8tW6uKGW9g61qEkG3bMqQPQWYEf5X3J",
+		wantCode: http.StatusBadRequest,
+	}, {
+		name:                 "error getting raw tx from dcrd client",
+		getRawTransactionErr: errors.New("get raw transaction error"),
+		wantCode:             http.StatusInternalServerError,
+	}, {
+		name:             "error getting can vote from dcrd client",
+		canTicketVoteErr: errors.New("can ticket vote error"),
+		wantCode:         http.StatusInternalServerError,
+	}, {
+		name:             "ticket can't vote",
+		canTicketNotVote: true,
+		wantCode:         http.StatusBadRequest,
+	}, {
+		name:      "bad sig",
+		addr:      testAddr,
+		deformSig: "x",
+		wantCode:  http.StatusInternalServerError,
+	}, {
+		name:          "hist at max",
+		addr:          testAddr,
+		setMaxHistory: true,
+		wantCode:      http.StatusBadRequest,
+	}, {
+		name:        "new ticket nothing to set",
+		isNewTicket: true,
+		wantCode:    http.StatusBadRequest,
+	}}
+
+	for _, test := range tests {
+		tTicket := exampleTicket()
+		if !test.isNewTicket {
+			if err := db.InsertNewTicket(tTicket); err != nil {
+				t.Fatalf("%q: unable to insert ticket: %v", test.name, err)
+			}
+		}
+
+		cfg.VspClosed = test.vspClosed
+
+		tNode := &testNode{
+			canTicketVote:        !test.canTicketNotVote,
+			canTicketVoteErr:     test.canTicketVoteErr,
+			getRawTransactionErr: test.getRawTransactionErr,
+		}
+		var w *httptest.ResponseRecorder
+		request := func() {
+			w = httptest.NewRecorder()
+			c, r := gin.CreateTestContext(w)
+
+			req := &SetAltSigRequest{
+				Timestamp:     time.Now().Unix(),
+				TicketHash:    tTicket.Hash,
+				TicketHex:     randString(504, hexCharset),
+				ParentHex:     randString(504, hexCharset),
+				AltPubAddress: test.addr,
+			}
+			b, err := json.Marshal(req)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			handle := func(c *gin.Context) {
+				c.Set("Ticket", tTicket)
+				c.Set("KnownTicket", !test.isNewTicket)
+				c.Set("CommitmentAddress", tTicket.Hash)
+				c.Set("SigningAddress", tTicket.AltSigAddress)
+				c.Set("DcrdClient", tNode)
+				c.Set("RequestBytes", b[test.deformReq:])
+				setAltSig(c)
+			}
+
+			r.POST("/", handle)
+
+			c.Request, err = http.NewRequest(http.MethodPost, "/", nil)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			reqSig := randString(504, hexCharset)
+			c.Request.Header.Set("VSP-Client-Signature", reqSig+test.deformSig)
+
+			r.ServeHTTP(w, c.Request)
+		}
+
+		// Max history is currently one.
+		if test.setMaxHistory {
+			for i := 0; i < 1; i++ {
+				request()
+			}
+		}
+		request()
+
+		resp := w.Result()
+		body, _ := ioutil.ReadAll(resp.Body)
+		fmt.Printf("\nThe response body for %q: %s\n\n", test.name, string(body))
```

Comment:
Debug line slipped through.

Reply:
I thought this was alright to leave in tests, as you can see the error going to the user, but I'll remove.